pkg/csi/service/fcd/controller.go

```diff
@@ -224,15 +225,101 @@ func (c *controller) ControllerPublishVolume(
 	req *csi.ControllerPublishVolumeRequest) (
 	*csi.ControllerPublishVolumeResponse, error) {
 
-	return nil, nil
+	//check for required parameters
+	if len(req.VolumeId) == 0 {
+		msg := fmt.Sprintf("Volume ID is a required parameter.")
+		log.Errorf(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	} else if len(req.NodeId) == 0 {
+		msg := fmt.Sprintf("Node ID is a required parameter.")
+		log.Errorf(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+
+	discoveryInfo, err := c.connMgr.WhichVCandDCByFCDId(ctx, req.VolumeId)
+	if err != nil {
+		msg := fmt.Sprintf("WhichVCandDCByFCDId(%s) failed. Err: %v", req.VolumeId, err)
+		log.Errorf(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+
+	fcd := discoveryInfo.FCDInfo
+
+	vm, err := discoveryInfo.DataCenter.GetVMByDNSName(ctx, req.NodeId)
+	if err != nil {
+		log.Errorf("GetVMByDNSName(%s) failed. Err: %v", req.NodeId, err)
+		return nil, err
+	}
+
+	filePath := fcd.Config.Backing.(*types.BaseConfigInfoDiskFileBackingInfo).FilePath
+	options := &vclib.VolumeOptions{SCSIControllerType: vclib.PVSCSIControllerType}
+	diskUUID, err := vm.AttachDisk(ctx, filePath, options)
```

Comment:
Should this be idempotent?  Can check if the disk is attached like so:
```go
                devices, err := vm.Device(ctx)
		disk := devices.SelectByBackingInfo(&types.VirtualDiskFlatVer2BackingInfo{
			VirtualDeviceFileBackingInfo: types.VirtualDeviceFileBackingInfo{
				FileName: filePath,
			},
		})
                if len(disk) != 0 { // disk is already attached }
```

Reply:
Yes, this should be idempotent.

Reply:
The idempotency takes place inside AttachDisk:
`	// If disk is already attached, return the disk UUID
	if attached {
		diskUUID, _ := vm.Datacenter.GetVirtualDiskPage83Data(ctx, vmDiskPath)
		return diskUUID, nil
	}`

Reply:
Could swear I checked first, but yeah can clearly see AttachDisk is idempotent.  And in looking a second time, I noticed there are VirtualMachine AttachDisk/DetachDisk API methods that were added in 6.5.  This may simplify things a bit, but probably functionally the same - I'll open an issue we can look closer later.